src/main/java/com/openlattice/chronicle/services/message/MessageService.kt

```diff
@@ -30,11 +33,68 @@ open class MessageService(
 ) : MessageManager {
 
     companion object {
-        protected val logger = LoggerFactory.getLogger(MessageService::class.java)
+        protected val logger: Logger = LoggerFactory.getLogger(MessageService::class.java)
+        private const val MAX_FAILURE_ATTEMPTS = 5
+
+    }
+
+    open fun getSearchTermString(propertyTypeId: UUID, searchString: String): String {
+        return "entity.$propertyTypeId:\"$searchString\""
+    }
+
+    @Throws(ExecutionException::class)
+    override fun trackUndeliveredMessage(organizationId: UUID, messageSid: String) {
+        try {
+            val apiClient = apiCacheManager.prodApiClientCache[ApiClient::class.java]
+            val dataApi = apiClient.dataApi
+            val searchApi = apiClient.searchApi
+            val coreAppConfig = entitySetIdsManager.getChronicleAppConfig(organizationId)
+
+            // entity set ids
+            val messageESID = coreAppConfig.messagesEntitySetId
+            val searchTerm = SearchTerm( getSearchTermString(
+                edmCacheManager.getPropertyTypeId(EdmConstants.OL_ID_FQN),
+                messageSid
+            ), 0, 1)
+
+            var numAttempts = 0
+            var entity: Map<FullQualifiedName, Set<Any>>? = null
+            while (entity == null && numAttempts <= MAX_FAILURE_ATTEMPTS) {
+                val searchResult = searchApi.executeEntitySetDataQuery( messageESID, searchTerm )
+                if (searchResult != null && searchResult.hits.isNotEmpty()) {
+                    entity = searchResult.hits[0]
+                }
+                try {
+                    Thread.sleep(1000) // if message sid not indexed in elasticsearch yet, wait a second and retry
+                } catch (e: InterruptedException) {
+                }
+                numAttempts++
+            }
+            if (entity != null) {
+                val entityKeyId = UUID.fromString(
+                        entity[EdmConstants.OL_EKID]!!.iterator().next().toString()
```

Comment:
no change required
`entity[EdmConstants.OL_EKID]!!` can be replaced with `entity.getValue(EdmConstants.OL_EKID) `